File: packages/rayban/modules/expo-meta-camera/android/ExpoMetaCameraModule.kt
package expo.modules.metacamera

import android.graphics.Bitmap
import android.util.Base64
import com.meta.wearable.dat.camera.Stream
import com.meta.wearable.dat.camera.addStream
import com.meta.wearable.dat.camera.types.PhotoData
import com.meta.wearable.dat.camera.types.StreamConfiguration
import com.meta.wearable.dat.camera.types.StreamError
import com.meta.wearable.dat.camera.types.StreamState
import com.meta.wearable.dat.camera.types.VideoFrame
import com.meta.wearable.dat.camera.types.VideoQuality
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.selectors.AutoDeviceSelector
import com.meta.wearable.dat.core.selectors.DeviceSelector
import com.meta.wearable.dat.core.session.DeviceSession
import com.meta.wearable.dat.core.session.DeviceSessionState
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeoutOrNull

/**
 * Android Meta DAT SDK camera integration for Ray-Ban Meta and Meta Ray-Ban
 * Display glasses.
 *
 * Architecture mirrors the iOS module:
 *  - `Wearables.initialize(ctx)` is invoked once at module load. Idempotent on the
 *    Android side as long as it's called from the same process.
 *  - `startStream`: creates a `DeviceSession` via `AutoDeviceSelector`, waits for
 *    `DeviceSessionState.STARTED`, then `addStream(...)` with `MEDIUM` quality at
 *    24 fps. Latest VideoFrame's converted bitmap is cached in memory.
 *  - `captureFrame`: returns a JPEG of the most recent cached frame. If no frame
 *    has arrived yet, calls `stream.capturePhoto()` and waits for the result.
 *  - `stopStream`: stops stream + session, cancels coroutines.
 *
 * Glasses-button taps: same indirect signal as iOS. The wearer tapping the side
 * touchpad pauses the camera session (DeviceSessionState.PAUSED); the SDK
 * auto-resumes it. We treat each PAUSED → STARTED edge as an `onTap` event,
 * skipping the very first STARTED transition (initial connection).
 */
class ExpoMetaCameraModule : Module() {
  // MARK: - Threading

  private val moduleScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

  // MARK: - State

  private var deviceSession: DeviceSession? = null
  private var stream: Stream? = null
  private var videoJob: Job? = null
  private var stateJob: Job? = null
  private var errorJob: Job? = null
  private var sessionStateJob: Job? = null

  /// Cached most-recent frame as raw JPEG bytes. Refreshed on each video frame
  /// callback (downsample + JPEG-encode happens off-Main).
  @Volatile private var latestJpegBytes: ByteArray? = null
  @Volatile private var latestWidth: Int = 0
  @Volatile private var latestHeight: Int = 0
  private val frameLock = Mutex()

  /// Pending photo waiters drained when photoData arrives via stream events.
  private val photoFlow = MutableSharedFlow<PhotoData>(replay = 0, extraBufferCapacity = 4)

  /// Tap-detection state. The first STARTED edge is the initial connection;
  /// only PAUSED → STARTED edges thereafter are wearer taps.
  @Volatile private var sessionHasStartedBefore: Boolean = false
  @Volatile private var lastSessionState: DeviceSessionState = DeviceSessionState.STOPPED
  @Volatile private var tapSequence: Int = 0

  private val deviceSelector: DeviceSelector by lazy { AutoDeviceSelector() }

  // MARK: - Module definition

  override fun definition() = ModuleDefinition {
    Name("ExpoMetaCameraModule")

    Events("onCameraState", "onCameraError", "onTap")

    OnCreate {
      val ctx = appContext.reactContext
      if (ctx != null) {
        try {
          Wearables.initialize(ctx)
        } catch (_: Throwable) {
          // Already initialized — fine.
        }
      }
    }

    AsyncFunction("startStream") { promise: Promise ->
      moduleScope.launch { startStreamAsync(promise) }
    }

    AsyncFunction("stopStream") { promise: Promise ->
      moduleScope.launch {
        stopStreamAsync()
        promise.resolve(null)
      }
    }

    AsyncFunction("captureFrame") { promise: Promise ->
      moduleScope.launch { captureFrameAsync(promise) }
    }

    OnDestroy {
      moduleScope.launch { stopStreamAsync() }
    }
  }

  // MARK: - Stream lifecycle

  private suspend fun startStreamAsync(promise: Promise) {
    if (stream != null) {
      sendEvent("onCameraState", mapOf("state" to "streaming"))
      promise.resolve(null)
      return
    }

    sendEvent("onCameraState", mapOf("state" to "connecting"))

    // Permission gate. The host app is responsible for the Meta AI registration
    // flow (Wearables.startRegistration) AND for prompting the wearer when
    // permission isn't yet granted (it requires an Activity-result contract,
    // which this module can't run from). We only check status here; if not
    // Granted, the host must drive the request and call startStream again.
    val permission = com.meta.wearable.dat.core.types.Permission.CAMERA
    var granted = false
    Wearables.checkPermissionStatus(permission)
        .onSuccess { status ->
          granted = status == com.meta.wearable.dat.core.types.PermissionStatus.Granted
        }
        .onFailure { error, _ ->
          sendEvent(
              "onCameraError",
              mapOf("code" to "permission-denied", "message" to "Permission check failed: ${error.description}")
          )
        }
    if (!granted) {
      sendEvent(
          "onCameraError",
          mapOf("code" to "permission-denied", "message" to "Camera permission not granted (host app must call Wearables.startRegistration / RequestPermissionContract)")
      )
      sendEvent("onCameraState", mapOf("state" to "disconnected"))
      promise.resolve(null)
      return
    }

    // Create the device session.
    var newSession: DeviceSession? = null
    Wearables.createSession(deviceSelector)
        .onSuccess { newSession = it }
        .onFailure { error, _ ->
          sendEvent(
              "onCameraError",
              mapOf("code" to "sdk-unavailable", "message" to "createSession failed: ${error.description}")
          )
          sendEvent("onCameraState", mapOf("state" to "error"))
        }
    val session = newSession ?: run { promise.resolve(null); return }
    deviceSession = session

    // Spawn the long-running session-state observer (drives onTap detection).
    sessionStateJob =
        moduleScope.launch {
          session.state.collect { handleSessionStateTransition(it) }
        }

    session.start()

    // Wait for the session to reach STARTED, with a 10 s timeout.
    val started =
        withTimeoutOrNull(10_000) {
          session.state.first { it == DeviceSessionState.STARTED }
        } != null
    if (!started) {
      session.stop()
      deviceSession = null
      sessionStateJob?.cancel()
      sessionStateJob = null
      sendEvent(
          "onCameraError",
          mapOf("code" to "connection-lost", "message" to "Device session did not reach STARTED")
      )
      sendEvent("onCameraState", mapOf("state" to "error"))
      promise.resolve(null)
      return
    }

    // Add a stream capability.
    var newStream: Stream? = null
    session
        .addStream(StreamConfiguration(videoQuality = VideoQuality.MEDIUM, frameRate = 24))
        .onSuccess { newStream = it }
        .onFailure { error, _ ->
          sendEvent(
              "onCameraError",
              mapOf("code" to "sdk-unavailable", "message" to "addStream failed: ${error.description}")
          )
        }
    val s = newStream ?: run { promise.resolve(null); return }
    stream = s
    setupStreamListeners(s)
    s.start()
    promise.resolve(null)
  }

  private suspend fun stopStreamAsync() {
    videoJob?.cancel()
    stateJob?.cancel()
    errorJob?.cancel()
    sessionStateJob?.cancel()
    videoJob = null
    stateJob = null
    errorJob = null
    sessionStateJob = null

    stream?.stop()
    stream = null
    deviceSession?.stop()
    deviceSession = null

    sessionHasStartedBefore = false
    lastSessionState = DeviceSessionState.STOPPED

    frameLock.withLock {
      latestJpegBytes = null
      latestWidth = 0
      latestHeight = 0
    }

    sendEvent("onCameraState", mapOf("state" to "disconnected"))
  }

  // MARK: - Frame capture

  private suspend fun captureFrameAsync(promise: Promise) {
    // Fast path: cached JPEG.
    val cached: Triple<ByteArray, Int, Int>? =
        frameLock.withLock {
          val bytes = latestJpegBytes
          if (bytes != null) Triple(bytes, latestWidth, latestHeight) else null
        }
    if (cached != null) {
      promise.resolve(
          mapOf(
              "data" to Base64.encodeToString(cached.first, Base64.NO_WRAP),
              "width" to cached.second,
              "height" to cached.third
          )
      )
      return
    }

    // Slow path: trigger photo capture.
    val activeStream = stream
    if (activeStream == null) {
      promise.reject(
          "CAMERA_NOT_STREAMING",
          "captureFrame called before stream is active",
          null
      )
      return
    }

    val photoData =
        withTimeoutOrNull(5_000) {
          // Issue capture in parallel with the wait for photoFlow to reduce latency.
          val captureLaunch = moduleScope.launch { activeStream.capturePhoto() }
          val received = photoFlow.first()
          captureLaunch.join()
          received
        }

    if (photoData == null) {
      promise.reject(
          "CAPTURE_TIMEOUT",
          "capturePhoto did not deliver a photo within 5s",
          null
      )
      return
    }

    val (jpegBytes, w, h) = photoToJpegBytes(photoData)
    promise.resolve(
        mapOf(
            "data" to Base64.encodeToString(jpegBytes, Base64.NO_WRAP),
            "width" to w,
            "height" to h
        )
    )
  }

  /**
   * Convert a `PhotoData` into JPEG bytes. The SDK delivers either a Bitmap or
   * raw HEIC; we always re-encode to JPEG so the LLM clients receive a uniform
   * format regardless of platform.
   */
  private fun photoToJpegBytes(photo: PhotoData): Triple<ByteArray, Int, Int> {
    return when (photo) {
      is PhotoData.Bitmap -> {
        val bmp = photo.bitmap
        val out = ByteArrayOutputStream()
        bmp.compress(Bitmap.CompressFormat.JPEG, 85, out)
        Triple(out.toByteArray(), bmp.width, bmp.height)
      }
      is PhotoData.HEIC -> {
        // Decode HEIC → Bitmap → JPEG. BitmapFactory handles HEIC on API 28+.
        val byteArray = ByteArray(photo.data.remaining())
        photo.data.get(byteArray)
        val bmp = android.graphics.BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
        val out = ByteArrayOutputStream()
        bmp.compress(Bitmap.CompressFormat.JPEG, 85, out)
        Triple(out.toByteArray(), bmp.width, bmp.height)
      }
    }
  }

  // MARK: - Stream listeners

  private fun setupStreamListeners(s: Stream) {
    videoJob =
        moduleScope.launch(Dispatchers.Default) {
          s.videoStream.collect { handleVideoFrame(it) }
        }

    stateJob =
        moduleScope.launch {
          s.state.collect { state ->
            val mapped =
                when (state) {
                  StreamState.STREAMING -> "streaming"
                  StreamState.CLOSED -> "disconnected"
                  StreamState.STARTING, StreamState.STOPPING, StreamState.PAUSED -> "connecting"
                  else -> "connecting"
                }
            sendEvent("onCameraState", mapOf("state" to mapped))
          }
        }

    errorJob =
        moduleScope.launch {
          s.errorStream.collect { error ->
            if (error == StreamError.STREAM_ERROR) return@collect
            sendEvent(
                "onCameraError",
                mapOf("code" to "connection-lost", "message" to error.description)
            )
          }
        }

    // photoDataPublisher equivalent: collect from the stream's photo flow if
    // exposed. The Android sample wires capturePhoto via a Result<PhotoData>
    // return value, so we don't need a separate listener — the photoFlow is
    // emitted from captureFrameAsync directly when capturePhoto resolves.
  }

  /// Convert an I420 video frame into a JPEG bitmap and cache it.
  /// Runs on Dispatchers.Default — the conversion is CPU-intensive and must
  /// not block the main thread.
  private suspend fun handleVideoFrame(frame: VideoFrame) {
    val bitmap = yuvI420ToBitmap(frame.buffer, frame.width, frame.height) ?: return
    val out = ByteArrayOutputStream()
    bitmap.compress(Bitmap.CompressFormat.JPEG, 75, out)
    val bytes = out.toByteArray()
    frameLock.withLock {
      latestJpegBytes = bytes
      latestWidth = frame.width
      latestHeight = frame.height
    }
  }

  /**
   * Minimal I420 → ARGB conversion. The Android sample app uses a more
   * optimised converter (`YuvToBitmapConverter` with IntArray pooling); for
   * the rayban use case we capture frames infrequently so a simpler
   * implementation is acceptable.
   *
   * `buffer` layout is I420 planar: Y plane (width * height bytes), then U
   * plane (width/2 * height/2 bytes), then V plane (width/2 * height/2 bytes).
   */
  private fun yuvI420ToBitmap(buffer: ByteBuffer, width: Int, height: Int): Bitmap? {
    if (width <= 0 || height <= 0) return null
    val frameSize = width * height
    val chromaSize = frameSize / 4

    val src = ByteArray(buffer.remaining())
    buffer.get(src)

    if (src.size < frameSize + 2 * chromaSize) return null

    val argb = IntArray(frameSize)
    val uOffset = frameSize
    val vOffset = frameSize + chromaSize

    for (y in 0 until height) {
      for (x in 0 until width) {
        val yIndex = y * width + x
        val uvIndex = (y / 2) * (width / 2) + (x / 2)
        val yVal = (src[yIndex].toInt() and 0xFF) - 16
        val uVal = (src[uOffset + uvIndex].toInt() and 0xFF) - 128
        val vVal = (src[vOffset + uvIndex].toInt() and 0xFF) - 128

        val yScale = 1.164f * yVal
        var r = (yScale + 1.596f * vVal).toInt()
        var g = (yScale - 0.392f * uVal - 0.813f * vVal).toInt()
        var b = (yScale + 2.017f * uVal).toInt()

        if (r < 0) r = 0 else if (r > 255) r = 255
        if (g < 0) g = 0 else if (g > 255) g = 255
        if (b < 0) b = 0 else if (b > 255) b = 255

        argb[yIndex] = (0xFF shl 24) or (r shl 16) or (g shl 8) or b
      }
    }

    return Bitmap.createBitmap(argb, width, height, Bitmap.Config.ARGB_8888)
  }

  // MARK: - Tap detection

  private fun handleSessionStateTransition(state: DeviceSessionState) {
    val prev = lastSessionState
    lastSessionState = state

    if (state == DeviceSessionState.STARTED) {
      if (prev == DeviceSessionState.PAUSED && sessionHasStartedBefore) {
        tapSequence += 1
        sendEvent("onTap", mapOf("sequence" to tapSequence))
      }
      sessionHasStartedBefore = true
    }
  }
}

File: packages/rayban/modules/expo-meta-camera/ios/ExpoMetaCameraModule.swift
import ExpoModulesCore
import MWDATCore
import MWDATCamera
import UIKit
import Foundation

/**
 * iOS Meta DAT SDK camera integration for Ray-Ban Meta and Meta Ray-Ban
 * Display glasses.
 *
 * Architecture:
 *  - `Wearables.configure()` is invoked once at module init (idempotent —
 *    if the host app already configured the SDK, this throws and is ignored).
 *  - `startStream`: creates a `DeviceSession` via `AutoDeviceSelector`,
 *    waits for `.started`, adds a `Stream` capability at low resolution
 *    (360×640) / 24 fps, and starts the stream. The most recent
 *    `VideoFrame` is cached in memory.
 *  - `captureFrame`: returns a JPEG of the most recent cached video frame
 *    (immediate, no round-trip). If no frame yet, requests a still photo
 *    via `stream.capturePhoto(.jpeg)` and waits for the
 *    `photoDataPublisher` to deliver.
 *  - `stopStream`: stops the stream and the underlying device session.
 *
 * Glasses-button taps: the DAT SDK has no explicit tap event API, but
 * tapping the Ray-Ban side touchpad pauses/resumes the camera session.
 * We surface that as `onTap` by watching `DeviceSession.state` for
 * `paused → started` transitions (the resume edge — equivalent to "tap
 * to capture"). The first `started` after startup is suppressed since
 * that's the initial connection, not a wearer gesture.
 *
 * Permission: the user must accept Meta AI's permission prompt for camera
 * access. The host app is responsible for completing the registration
 * flow (`Wearables.shared.startRegistration()`); this module only consumes
 * an already-registered session.
 */
public class ExpoMetaCameraModule: Module {
  // MARK: - State

  private var deviceSession: DeviceSession?
  private var stream: MWDATCamera.Stream?
  private var stateListenerToken: AnyListenerToken?
  private var videoFrameListenerToken: AnyListenerToken?
  private var errorListenerToken: AnyListenerToken?
  private var photoDataListenerToken: AnyListenerToken?
  private var sessionStateTask: Task<Void, Never>?

  /// Tracks paused→started edges. The first `.started` is the initial connection;
  /// only subsequent `.paused → .started` edges are treated as wearer taps.
  private var sessionHasStartedBefore: Bool = false
  private var lastSessionState: DeviceSessionState = .stopped
  private var tapSequence: Int = 0

  /// Cached latest video frame so `captureFrame()` can return synchronously.
  private var latestFrame: UIImage?
  private let frameLock = NSLock()

  /// Pending continuations waiting for a photo to arrive after `capturePhoto`.
  private var pendingPhotoContinuations: [CheckedContinuation<Data, Error>] = []
  private let photoLock = NSLock()

  // MARK: - Module definition

  public func definition() -> ModuleDefinition {
    Name("ExpoMetaCameraModule")

    Events("onCameraState", "onCameraError", "onTap")

    OnCreate {
      // Configure the SDK once at module load. The host app may have already
      // called this — `Wearables.configure()` throws on a second call, which
      // we treat as harmless.
      _ = try? Wearables.configure()
    }

    AsyncFunction("startStream") { [weak self] (promise: Promise) in
      guard let self else { promise.resolve(); return }
      Task { @MainActor in
        await self.startStreamAsync(promise: promise)
      }
    }

    AsyncFunction("stopStream") { [weak self] (promise: Promise) in
      guard let self else { promise.resolve(); return }
      Task { @MainActor in
        await self.stopStreamAsync()
        promise.resolve()
      }
    }

    AsyncFunction("captureFrame") { [weak self] (promise: Promise) in
      guard let self else {
        promise.reject("CAMERA_UNAVAILABLE", "module deallocated")
        return
      }
      Task { @MainActor in
        await self.captureFrameAsync(promise: promise)
      }
    }

    OnDestroy {
      Task { @MainActor [weak self] in
        await self?.stopStreamAsync()
      }
    }
  }

  // MARK: - Stream lifecycle

  @MainActor
  private func startStreamAsync(promise: Promise) async {
    if stream != nil {
      // Already streaming — emit current state and resolve.
      sendEvent("onCameraState", ["state": "streaming"])
      promise.resolve()
      return
    }

    sendEvent("onCameraState", ["state": "connecting"])

    // Permission gate.
    do {
      let wearables = Wearables.shared
      var status = try await wearables.checkPermissionStatus(.camera)
      if status != .granted {
        status = try await wearables.requestPermission(.camera)
      }
      if status != .granted {
        sendEvent("onCameraError", [
          "code": "permission-denied",
          "message": "Camera permission was not granted",
        ])
        sendEvent("onCameraState", ["state": "disconnected"])
        promise.resolve()
        return
      }
    } catch {
      sendEvent("onCameraError", [
        "code": "sdk-unavailable",
        "message": "Permission check failed: \(error.localizedDescription)",
      ])
      sendEvent("onCameraState", ["state": "error"])
      promise.resolve()
      return
    }

    // Create a device session.
    let session: DeviceSession
    do {
      let wearables = Wearables.shared
      let selector = AutoDeviceSelector(wearables: wearables)
      session = try wearables.createSession(deviceSelector: selector)
      try session.start()

      // Wait for the session to reach .started, with a 10 s timeout.
      let deadline = Date().addingTimeInterval(10.0)
      var started = session.state == .started
      if !started {
        for await state in session.stateStream() {
          if state == .started { started = true; break }
          if state == .stopped { break }
          if Date() > deadline { break }
        }
      }
      guard started else {
        session.stop()
        sendEvent("onCameraError", [
          "code": "connection-lost",
          "message": "Device session did not reach .started",
        ])
        sendEvent("onCameraState", ["state": "error"])
        promise.resolve()
        return
      }
      deviceSession = session
      sessionHasStartedBefore = true
      lastSessionState = .started
      // Spawn a long-running observer that emits onTap on each paused→started edge.
      sessionStateTask = Task { [weak self] in
        guard let self else { return }
        for await state in session.stateStream() {
          await self.handleSessionStateTransition(state)
        }
      }
    } catch {
      sendEvent("onCameraError", [
        "code": "sdk-unavailable",
        "message": "createSession failed: \(error.localizedDescription)",
      ])
      sendEvent("onCameraState", ["state": "error"])
      promise.resolve()
      return
    }

    // Add a stream capability. Low resolution + 24 fps gives the smallest
    // BT bandwidth footprint; we don't need pristine quality for LLM input.
    let config = StreamConfiguration(
      videoCodec: .raw,
      resolution: .low,
      frameRate: 24
    )
    guard let newStream = try? session.addStream(config: config) else {
      sendEvent("onCameraError", [
        "code": "sdk-unavailable",
        "message": "addStream returned nil",
      ])
      sendEvent("onCameraState", ["state": "error"])
      promise.resolve()
      return
    }
    stream = newStream
    setupStreamListeners(for: newStream)
    await newStream.start()
    promise.resolve()
  }

  @MainActor
  private func stopStreamAsync() async {
    sessionStateTask?.cancel()
    sessionStateTask = nil
    sessionHasStartedBefore = false
    lastSessionState = .stopped

    if let activeStream = stream {
      stream = nil
      clearStreamListeners()
      await activeStream.stop()
    }
    if let activeSession = deviceSession {
      deviceSession = nil
      activeSession.stop()
    }
    frameLock.lock()
    latestFrame = nil
    frameLock.unlock()

    // Reject any pending photo waiters.
    photoLock.lock()
    let waiters = pendingPhotoContinuations
    pendingPhotoContinuations.removeAll()
    photoLock.unlock()
    for c in waiters {
      c.resume(throwing: CameraModuleError.streamStopped)
    }

    sendEvent("onCameraState", ["state": "disconnected"])
  }

  // MARK: - Frame capture

  @MainActor
  private func captureFrameAsync(promise: Promise) async {
    // Fast path: return the most recent cached frame.
    frameLock.lock()
    let cached = latestFrame
    frameLock.unlock()
    if let image = cached, let data = image.jpegData(compressionQuality: 0.85) {
      promise.resolve([
        "data": data.base64EncodedString(),
        "width": Int(image.size.width),
        "height": Int(image.size.height),
      ])
      return
    }

    // Slow path: trigger a still capture via the SDK's photo path.
    guard let activeStream = stream else {
      promise.reject("CAMERA_NOT_STREAMING", "captureFrame called before stream is active")
      return
    }

    do {
      let data = try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Data, Error>) in
        photoLock.lock()
        pendingPhotoContinuations.append(cont)
        photoLock.unlock()
        let success = activeStream.capturePhoto(format: .jpeg)
        if !success {
          // Roll back: pop the continuation we just pushed and reject it.
          photoLock.lock()
          if let last = pendingPhotoContinuations.last {
            pendingPhotoContinuations.removeLast()
            photoLock.unlock()
            last.resume(throwing: CameraModuleError.captureFailed)
          } else {
            photoLock.unlock()
          }
        }
      }
      // Decode dimensions from JPEG header.
      let img = UIImage(data: data)
      promise.resolve([
        "data": data.base64EncodedString(),
        "width": Int(img?.size.width ?? 0),
        "height": Int(img?.size.height ?? 0),
      ])
    } catch {
      promise.reject("CAPTURE_FAILED", error.localizedDescription)
    }
  }

  // MARK: - Stream listeners

  private func setupStreamListeners(for s: MWDATCamera.Stream) {
    stateListenerToken = s.statePublisher.listen { [weak self] state in
      Task { @MainActor in self?.handleStateChange(state) }
    }
    videoFrameListenerToken = s.videoFramePublisher.listen { [weak self] frame in
      // Frame callbacks fire on a background queue; lock-only critical
      // section, no UI work here.
      guard let self else { return }
      if let image = frame.makeUIImage() {
        self.frameLock.lock()
        self.latestFrame = image
        self.frameLock.unlock()
      }
    }
    errorListenerToken = s.errorPublisher.listen { [weak self] error in
      Task { @MainActor in self?.handleStreamError(error) }
    }
    photoDataListenerToken = s.photoDataPublisher.listen { [weak self] data in
      guard let self else { return }
      self.photoLock.lock()
      let waiters = self.pendingPhotoContinuations
      self.pendingPhotoContinuations.removeAll()
      self.photoLock.unlock()
      for c in waiters {
        c.resume(returning: data.data)
      }
    }
  }

  private func clearStreamListeners() {
    stateListenerToken = nil
    videoFrameListenerToken = nil
    errorListenerToken = nil
    photoDataListenerToken = nil
  }

  @MainActor
  private func handleStateChange(_ state: StreamState) {
    let mapped: String
    switch state {
    case .streaming: mapped = "streaming"
    case .stopped: mapped = "disconnected"
    case .stopping, .starting, .waitingForDevice, .paused: mapped = "connecting"
    @unknown default: mapped = "connecting"
    }
    sendEvent("onCameraState", ["state": mapped])
  }

  @MainActor
  private func handleStreamError(_ error: StreamError) {
    sendEvent("onCameraError", [
      "code": "connection-lost",
      "message": error.localizedDescription,
    ])
  }

  /// Convert DeviceSession state transitions into `onTap` events.
  /// The wearer's side-touchpad gesture pauses the camera session; the
  /// SDK then auto-resumes it. The paused → started edge is therefore a
  /// reliable proxy for "tap to capture".
  @MainActor
  private func handleSessionStateTransition(_ state: DeviceSessionState) {
    let prev = lastSessionState
    lastSessionState = state

    if state == .started {
      if prev == .paused && sessionHasStartedBefore {
        tapSequence += 1
        sendEvent("onTap", ["sequence": tapSequence])
      }
      sessionHasStartedBefore = true
    }
  }
}

// MARK: - Errors

private enum CameraModuleError: LocalizedError {
  case streamStopped
  case captureFailed

  var errorDescription: String? {
    switch self {
    case .streamStopped: return "Stream was stopped before capture completed"
    case .captureFailed: return "capturePhoto returned false"
    }
  }
}
